qutebrowser/misc/mhtml.py

```diff
@@ -0,0 +1,411 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2015 Daniel Schadt
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utils for writing a MHTML file."""
+
+import functools
+import io
+import os
+import re
+
+from collections import namedtuple
+from base64 import b64encode
+from uuid import uuid4
+from email import policy
+from email.mime.multipart import MIMEMultipart
+from email.generator import BytesGenerator
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.utils import log, objreg, message
+
+
+_File = namedtuple("_File",
+                   "content content_type content_location transfer_encoding")
+
+
+_CSS_URL_PATTERNS = [re.compile(x) for x in [
+    rb"@import '(?P<url>[^']+)'",
+    rb'@import "(?P<url>[^"]+)"',
+    rb'''url\((?P<url>[^'"][^)]*)\)''',
+    rb'url\("(?P<url>[^"]+)"\)',
+    rb"url\('(?P<url>[^']+)'\)",
+]]
+
+
+def _get_css_imports(data):
+    """Return all assets that are referenced in the given CSS document.
+
+    The returned URLs are relative to the stylesheet's URL.
+
+    Args:
+        data: The content of the stylesheet to scan.
+    """
+    if isinstance(data, str):
+        data = data.encode("utf-8")
+    urls = []
+    for pattern in _CSS_URL_PATTERNS:
+        for match in pattern.finditer(data):
+            url = match.group("url")
+            if url:
+                urls.append(url)
+    return urls
+
+
+MHTMLPolicy = policy.default.clone(linesep="\r\n", max_line_length=0)
+
+
+def _chunked_base64(data, maxlen=76, linesep=b"\r\n"):
+    """Just like b64encode, except that it breaks long lines.
+
+    Args:
+        maxlen: Maximum length of a line, not including the line separator.
+        linesep: Line separator to use as bytes.
+    """
+    encoded = b64encode(data)
+    result = []
+    for i in range(0, len(encoded), maxlen):
+        result.append(encoded[i:i + maxlen])
+    return linesep.join(result)
+
+
+def _rn_quopri(data):
+    """Return a quoted-printable representation of data."""
+    # See RFC 2045 https://tools.ietf.org/html/rfc2045#section-6.7
+    # The stdlib version in the quopri module has inconsistencies with line
+    # endings and breaks up character escapes over multiple lines, which isn't
+    # understood by qute and leads to jumbled text
```

Comment:
Not really, some of the inconsistencies can be fixed by setting `quopri.b2a_qp = quopri.a2b_qp = None`, but that still doesn't produce the right output.

I've also noticed the [email.encoders.encode_quopri](https://docs.python.org/3.5/library/email.encoders.html#email.encoders.encode_quopri) and the (undocumented?) `email.quoprimime` module, so I'll maybe be able to use one of those instead.